### PR TITLE
Context Parallel Squashed MR for models dir

### PR DESCRIPTION
**Pull request by jomitchellnv**
Adds context parallelism support to ESM2 models inside `bionemo-recipes/models/esm2`

- Adds a comprehensive unit test that checks (1) Gradients (2) Loss and (3) Logits for NON-CP vs CP execution
- Adds a unit test that confirms that padded THD and unpadded THD return the same logits after the models forward pass
- Adds a Collator capable of computing CP_SHARDs for a given CP rank, which is amenable for downstream scattering to other GPUs. All in pre-fetch.

#### Usage

```python
TODO: Add code snippet
```

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Pre-submit Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass successfully

### CODE REVIEW

File: bionemo-recipes/models/esm2/src/esm/modeling_esm_te.py [hunk 1]
```diff
@@ -300,7 +306,6 @@ def __init__(self, config: NVEsmConfig, add_pooling_layer: bool = True):
             add_pooling_layer (bool): Whether to add a pooling layer.
         """
         super().__init__(config)
-        self.config = config
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/modeling_esm_te.py:303]: we dont need this right?

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 2]
```diff
@@ -269,6 +296,53 @@ def _pad_batch_to_multiple_of(self, batch):
...
+        """Initialize the MLMDataCollatorWithFlatteningCPAware.
+
+        Args:
+            collator: The collator to use for masking tokens.
+            cp_world_size: The size of the context parallelism group.
+        """
+        self.collator = collator
+        self.cp_world_size = cp_world_size
+
+    def __call__(self, features) -> List[dict[str, Any]]:
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/src/esm/collator.py]: just use `list[]`

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 3]
```diff
@@ -441,3 +515,108 @@ def _pt_pad_to_multiple_of(batch: dict[str, Any], pad_to_multiple_of: int, token
...
+# TODO(@jomitchell): Once this gets merged: https://github.com/NVIDIA/TransformerEngine/pull/2387
+# we can replace this with the one in TransformerEngine.
+def split_batch_by_cp_rank(
+    cu_seqlens_padded: torch.Tensor,
+    input_ids_padded: torch.Tensor,
+    labels_padded: torch.Tensor,
+    cp_group: torch.distributed.ProcessGroup = None,
+    qvk_format: str = "thd",
+    cp_rank: Optional[int] = None,
+    cp_world_size: Optional[int] = None,
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/src/esm/collator.py]: use `int | None`

File: bionemo-recipes/models/esm2/tests/test_cp.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+        check=False,
+        text=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        timeout=240,
+    )
+    if result.returncode != 0:
+        print(f"STDOUT:\n{result.stdout}")
+        print(f"STDERR:\n{result.stderr}")
+        pytest.fail(f"Command failed with exit code {result.returncode}")
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/tests/test_cp.py:193]: this file won't run at all on our 1-gpu CI nodes -- is it worth it at all to run a 1-GPU version just to make sure these functions all fire correctly?

File: bionemo-recipes/models/esm2/tests/test_cp.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+                    reconstructed_logits[chunk_start_in_seq:chunk_end_in_seq, :] = logits_list[1][
+                        cp_offset_rank1 : cp_offset_rank1 + chunk_size, :
+                    ]
+                    cp_offset_rank1 += chunk_size
+
+        assert reconstructed_logits.shape == logits_nondistributed_for_comparison.shape
+        torch.testing.assert_close(
+            reconstructed_logits,
+            logits_nondistributed_for_comparison,
+            atol=0.29,
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/tests/test_cp.py:355]: that's very specific 😟  -- is this just large enough to pass on your machine? can we get a bit of a safety margin for future library + hardware changes?

File: bionemo-recipes/models/esm2/tests/test_cp.py (hunk 4 above)
Comment (jomitchellnv) [reply bionemo-recipes/models/esm2/tests/test_cp.py:193]: nope -- don't run it on 1 gpu nodes, run it on 2 gpu nodes only

File: bionemo-recipes/models/esm2/tests/test_cp.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+                print(f"  Max absolute diff: {(grad_cp - grad_nondist).abs().max():.6f}")
+
+                # Check if they're close
+                try:
+                    torch.testing.assert_close(
+                        grad_cp, grad_nondist, atol=2e-3, rtol=1e-2, msg=f"Gradients don't match for {key}"
+                    )
+                    print("  ✓ Gradients match within tolerance")
+                except AssertionError as e:
+                    print(f"  ✗ Gradients differ: {e}")
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/tests/test_cp.py]: thanks cursor 😆  -- seriously though just do the guarded
```python
torch.testing.assert_close(
    grad_cp, grad_nondist, atol=2e-3, rtol=1e-2, msg=lambda x: f"Gradients don't match for {key}: {x}"
)
```
i'd drop all those prints() too if it were me

to be clear -- you _want_ this to fail if the grads don't match up. That's the whole point of the test

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 7]
```diff
@@ -269,6 +296,53 @@ def _pad_batch_to_multiple_of(self, batch):
         )
 
 
+class MLMDataCollatorWithFlatteningCPAware:
+    """A collator that is aware of context parallelism."""
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/src/esm/collator.py:300]: this is a good place to let cursor go to town, might be nice to get slightly more informative docstrings here. E.g., a quick overview of how TE expects input data for CP and how this needs to be used together with torch.distributed.scatter_object_list to send the data to each rank

File: bionemo-recipes/models/esm2/src/esm/collator.py (hunk 2 above)
Comment (jomitchellnv) [reply bionemo-recipes/models/esm2/src/esm/collator.py]: List > list. why tho

File: bionemo-recipes/models/esm2/tests/test_cp.py (hunk 5 above)
Comment (jomitchellnv) [reply bionemo-recipes/models/esm2/tests/test_cp.py:355]: My MAD=0.125 so *2 =0.25 + buffer = 0.29.

File: bionemo-recipes/models/esm2/src/esm/collator.py (hunk 3 above)
Comment (jomitchellnv) [reply bionemo-recipes/models/esm2/src/esm/collator.py]: like `cp_rank: int | None = None`? Ok

File: bionemo-recipes/models/esm2/tests/test_cp.py (hunk 6 above)
Comment (jomitchellnv) [reply bionemo-recipes/models/esm2/tests/test_cp.py]: yup -- done

File: bionemo-recipes/models/esm2/src/esm/collator.py (hunk 7 above)
Comment (jomitchellnv) [reply bionemo-recipes/models/esm2/src/esm/collator.py:300]: I planned on doing most of this inside the `recipe MR`: https://github.com/NVIDIA/bionemo-framework/pull/1338 because that's really where this is used, that MR is going to define the contract between the collator and the CPDataloader. I just had to copy that file into this MR so that I could use the `split_batch_by_cp_rank` part of it. So let's leave it to the next one.

**Review by pstjohn — APPROVED** [review]

File: bionemo-recipes/models/esm2/src/esm/modeling_esm_te.py (hunk 1 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/src/esm/modeling_esm_te.py:303]: Probably not if we don't access it

File: bionemo-recipes/models/esm2/src/esm/collator.py (hunk 2 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/src/esm/collator.py]: Saves u the import
